lib/sps/detail/Sps.cpp

```diff
@@ -0,0 +1,235 @@
+#include "../../AstroAccelerate/host_get_user_input.h"
+#include "../../AstroAccelerate/host_help.h"
+#include "../../AstroAccelerate/host_rfi.h"
+#include "../../AstroAccelerate/host_stratagy.h"
+#include "../../AstroAccelerate/host_write_file.h"
+
+#include "../../AstroAccelerate/params.h"
+
+#include "../../AstroAccelerate/host_debug.h"
+
+
+namespace ska {
+namespace astroaccelerate {
+namespace sps {
+
+template<typename SpsParameterType>
+Sps<SpsParameterType>::Sps(InputData &input_data,
+		 DedispersionPlan  &dedispersion_plan,
+		 UserInput &user_input)
+		 : _num_tchunks(dedispersion_plan.get_num_tchunks()),
+		 _range(dedispersion_plan.get_range()),
+		 _nchans(dedispersion_plan.get_nchans()),
+		 _maxshift(dedispersion_plan.get_maxshift()),
+		 _tsamp(dedispersion_plan.get_tsamp()),
+		 _max_ndms(dedispersion_plan.get_max_ndms()),
+		 _sigma_cutoff(user_input.get_sigma_cutoff()),
+		 _ndms(dedispersion_plan.get_ndms()),
+		 _dmshifts(dedispersion_plan.get_dmshifts()),
+		 _t_processed(dedispersion_plan.get_t_processed()),
+		 _dm_low(dedispersion_plan.get_dm_low()),
+		 _dm_high(dedispersion_plan.get_dm_high()),
+		 _dm_step(dedispersion_plan.get_dm_step()),
+		 _in_bin(user_input.get_in_bin()),
+		 _out_bin(user_input.get_out_bin()),
+		 _input_buffer(input_data.get_input_buffer())
+{
+	_gpu_input_size 	= 0;
+	_d_input 			= NULL;
+	_gpu_output_size 	= 0;
+	_d_output 			= NULL;
+}
+
+template<typename SpsParameterType>
+Sps<SpsParameterType>::~Sps()
+{
+	cudaFree(_d_input);
+	cudaFree(_d_output);
+}
+
+
+template<typename SpsParameterType>
+void Sps<SpsParameterType>::allocate_memory_cpu_output(DedispersionPlan const &dedispersion_plan,
+													   float ****output_buffer,
+													   size_t *output_size)
+{
+	*output_buffer = (float ***) malloc(dedispersion_plan.get_range() * sizeof(float **));
+	for (int i = 0; i < dedispersion_plan.get_range(); ++i)
+	{
+		int total_samps = 0;
+		for (int k = 0; k < dedispersion_plan.get_num_tchunks(); ++k)
+			total_samps += dedispersion_plan.get_t_processed()[i][k];
+		(*output_buffer)[i] = (float **) malloc(dedispersion_plan.get_ndms()[i] * sizeof(float *));
+		for (int j = 0; j < dedispersion_plan.get_ndms()[i]; ++j)
+		{
+			(*output_buffer)[i][j] = (float *) malloc(( total_samps ) * sizeof(float));
+		}
+		*output_size += ( total_samps ) * dedispersion_plan.get_ndms()[i] * sizeof(float);
+	}
+}
+
+template<typename SpsParameterType>
+void Sps<SpsParameterType>::allocate_memory_gpu(DedispersionPlan const &dedispersion_plan)
+{
+	int time_samps = dedispersion_plan.get_t_processed()[0][0] + dedispersion_plan.get_maxshift();
+	_gpu_input_size = time_samps * (size_t) dedispersion_plan.get_nchans() * sizeof(unsigned short);
+
+	printf("\n gpu inputsize: %d\n", (int)(_gpu_input_size/1024/1024));
+	cudaError_t rc1 = ( cudaMalloc((void **)&_d_input, _gpu_input_size) );
+	if (rc1 != cudaSuccess)
+	    printf("Could not allocate memory: %d", rc1);
```

Comment:
should either throw an exception or return an error code depending on your coding conventions.